hips/hip-00NN.md

```diff
@@ -0,0 +1,101 @@
+---
+hip: "00NN"
+title: "Support Git protocol for installing chart dependencies"
+authors: [ "Jeff Valore (@rally25rs)", "yxxhero <aiopsclub@163.com>", "Dominykas Blyžė <hello@dominykas.com>" ]
+created: "2021-10-05"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+Currently, Helm supports installing dependencies from various registries, however that requires the charts to be published there. There are use cases which are cumbersome when registries are involved, for example:
+
+- private charts for organizations/individuals who do not have the capacity to maintain the infrastructure required to run a private registry
+- testing work-in-progress charts in their umbrella charts
+
+## Motivation
+
+There are existing ways to achieve installation of charts without using a registry, however they are not very user-friendly and require additional tooling.
+
+- A Helm chart repository is effectively an `index.yaml` with links to downloads. Maintaining such a repository does create a burden of scripts and automation (e.g. Github Actions). This is not always feasible for smaller projects. It also does not really offer an obvious and readily available way of testing pre-releases.
+- For the testing use cases, charts can be packaged using `helm package`, however this does introduce manual steps and requires extra work to replicate in CI/CD scenarios.
+- There are several plugins available to solve this problem ([`aslafy-z/helm-git`](https://github.com/aslafy-z/helm-git), [`diwakar-s-maurya/helm-git`](https://github.com/diwakar-s-maurya/helm-git), [`sagansystems/helm-github`](https://github.com/sagansystems/helm-github)), however using plugins requires additional setup, which may not always be feasible (esp. in more complex team structures and cluster setups with advanced tooling, e.g. ArgoCD). An additional drawback is that the `Chart.yaml` does not provide a way to specify the plugin requirements, which leaves it up to the consumer to figure this out.
+
+## Rationale
+
+Installing dependencies from git is an established pattern in other ecosystems even when they are registry-based, e.g. npm (Node.js), pipenv (Python), bundler (Gem) have this option - it would make sense to have the behavior replicated in Helm.
+
+At least the npm and the pip ecosystems have already established a syntax of `vcs+protocol` for defining the dependency source, so it should be familiar to some users.
+
+One alternative to consider would be to exclude defining the vcs/protocol, similar to what Go does, esp. given that Helm is built using Go. This does limit the flexibility somewhat - while Go does allow adding a VCS qualifier at the end of the URL (allowing future support for other VCSs), it does not allow specifying the protocol, which means that the users might have to override the default protocol in their VCS configuration.
+
+## Specification
+
+The `Chart.yaml` should support the following format for `dependencies`:
```

Comment:
One curiosity I think we need to think about. Is for the "testing" use case, a user might want to refer to a git repository as a dependency. But they probably want to refer to a registry/OCI repository for non-testing.

The workflow of manually updating dependencies post-testing is possible of course. But it can be awkward. UX could be improved later too of course.

Reply:
We've definitely experienced this problem when using `npm`. In the end, we wrote tooling which does the dependency overrides in CI (via params and conventions), so that we don't have to commit things like this into git, as well as validators to fail release builds in case they find git based dependencies.

Would it make sense to print a warning during `helm dep build` / `helm dep up`? I don't think it's done for the `file://` based ones, although the use cases are probably different.

`helm package` could also pay attention to that and require a flag, although I'm not sure if the current implementation even validates what's under `chart/*` before doing the work? I just deleted a file inside the `tgz` under `charts` and it didn't flinch.

The "How to teach this" section does mention that the docs should explicitly recommend registries over git - probably a place for some notes there?

Reply:
> Would it make sense to print a warning during helm dep build / helm dep up?

I think this is more of a lint scenario. But curious to what others thoughts might be...

To me, it is not explicit that a git dependency can not be used for production. But that you probably should not (especially not a mutable reference). 

OTOH, we could decide to be cautious I think. e.g. `helm push/package` could refuse to publish a chart with a git dependency. And then folk can argue if that should be relaxed. Once we allow something, the horse has bolted, and we can't rescind that behavior (even if we later think it is a terrible idea).

Reply:
I agree that a linting rule should print a warning - adding that to the HIP: https://github.com/helm/community/pull/321/commits/c16a922d53efd9d2d47723246f8a37e0a475af87

I'm unconvinced about failing to package/push. I can see how a warning could be useful, but having to use a flag to publish might break people's CI scenarios? Is there an established way to solicit more feedback on this?